server/shared/models/query_builder.js

```diff
@@ -1,44 +1,140 @@
+    };
+  }
+  return queryObj;
+});
+
+/**
+ * Dependant on options adds between clause and values to query object
+ * @param   {Options} options options object
+ * @param   {QueryObject} queryObj
+ * @returns {QueryObject}
+ */
+const addBetweenClause = curry((options, queryObj) => {
+  const valuesOffset = queryObj.values.length + 1;
+
+  if (options.between) {
+    const betweenClause = `${
+      options.between.column
+    } BETWEEN  $${valuesOffset} AND $${valuesOffset + 1}`;
+    const joiner = options.where ? ` AND ` : ` WHERE `;
+    return {
+      text: `${queryObj.text} ${joiner} ${betweenClause}`,
+      values: [...queryObj.values, ...options.between.values],
+    };
+  }
+  return queryObj;
+});
+
+/**
+ * Dependant on options adds sort clause to query object
+ * @param   {Options} options options object
+ * @param   {QueryObject} queryObj
+ * @returns {QueryObject}
+ */
+const addSortClause = curry((options, queryObj) => {
+  if (options.sort) {
+    return {
+      text: `${queryObj.text} ORDER BY ${options.sort}`,
+      values: [...queryObj.values],
+    };
+  }
+  return queryObj;
+});
 
 /**
- * Generates WHERE clause to be appended onto query string,
- * as well as values to be substituted into parameterised query
- * @param   {Object} query      Specifies the constraint
- * @param   {Number} [offset=0] Offset for "$x" placeholders
- * @returns {[String, Any[]]}   2-element array of [query, values]
+ * Dependant on options adds pagination clause and values to query object
+ * @param   {Options} options options object
+ * @param   {QueryObject} queryObj
+ * @returns {QueryObject}
  */
-const whereClause = (query, offset = 0) =>
-  [
-    `WHERE ${Object.keys(query).map((k, i) => `${k}=$${i + 1 + offset}`).join(' AND ')}`,
-    Object.values(query),
-  ];
+const addPagination = curry((options, queryObj) => {
+  const valuesOffset = queryObj.values.length + 1;
+
+  if (options.pagination) {
+    const { offset } = options.pagination;
+
+    return {
+      text: `${queryObj.text} LIMIT 10 OFFSET $${valuesOffset}`,
+      values: [...queryObj.values, offset],
+    };
+  }
+  return queryObj;
+});
 
+const addReturning = curry((options, queryObj) => {
+  if (options.returning) {
+    return {
+      text: `${queryObj.text} RETURNING ${options.returning}`,
+      values: [...queryObj.values],
+    };
+  }
+  return queryObj;
+});
 
 /**
  * Generates SELECT query object with `text` and `value` fields
  * @param   {String}   table      Table name
  * @param   {String[]} columns    Columns to select
- * @param   {Object}   [where={}] Constraint object
- * @returns {Query}               { text: String, values: Array }
+ * @param   {Options}  options    options object
+ * @returns {QueryObject} queryObject
  */
-const selectQuery = (table, columns, where = {}) => {
+const selectQuery = (table, columns, options) => {
+  options.pagination && [`COUNT(*) OVER() AS full_count`].concat(columns); //eslint-disable-line
```

Comment:
This does nothing.

Reply:
😅, yeah I was struggling to think of the best way to do this with concat over my previous ternary. What do you suggest?

Reply:
const cols = (options.pagination ? 'COUNT(*) OVER() AS full_count' : []).concat(columns);
const base = `SELECT ${cols} FROM ${table}`;